### ISSUE TITLE
Error sfab_FabricatedSObject.FieldIsNotParentRelationshipException: The field Subscription__c.Organisation__c is not a parent relationship

### ISSUE THREAD

**Issue by cdonnelly-cms** [mattaddy/SObjectFabricator#21]
Hi,
Thank you for providing this library.  It sounds like a great way to improve unit test execution times.

I'm trying to use it for the first time and I'm getting the following runtime error:

    sfab_FabricatedSObject.FieldIsNotParentRelationshipException: The field Subscription__c.Organisation__c is not a parent relationship  
    Class.sfab_FabricatedSObject.checkFieldIsParentRelationship: line 534, column 1                                                       
    Class.sfab_FabricatedSObject.setDirectParent: line 389, column 1                                                                      
    Class.sfab_FabricatedSObject.setParent: line 240, column 1                                                                            
    Class.sfab_FabricatedSObject.set: line 112, column 1                                                                                  
    Class.sfab_FabricatedSObject: line 192, column 1                                                                                      
    Class.sfab_FabricatedSObject.<init>: line 58, column 1                                                                                
    Class.COMMS_subscription_TRIGGER_mc_TEST.after_update_organisation_subscription_happy: line 84, column 1 

My code is:

        Map<string,object> publicationGeneralMap = new Map<string,object>
            { 'Id'                  => 'Id-Pub-Gen-1'
            , 'Name'                => SETUP_NAME_PUBLICATION
            , 'RecordTypeId'        => toRecordTypeId(Schema.SObjectType.Publication__c, 'General_Publication')
            , 'Branch_Ownership__c' => SETUP_BRANCH_OWNERSHIP
            , 'Status__c'           => SETUP_PUBLICATION_STATUS
            , 'Email__c'            => true
            };
        sfab_FabricatedSObject fabPubGen = new sfab_FabricatedSObject( Publication__c.class, publicationGeneralMap );

        Map<string,object> accountGeneralMap = new Map<string,object>
            { 'Id'                    => 'Id-Acc-Gen-1'
            , 'Name'                  => SETUP_NAME_GENERAL_ACCOUNT
            , 'RecordTypeId'          => toRecordTypeId(Schema.SObjectType.Account, 'General_Account')
            , 'CMS_ID__c'             => SETUP_ORGANISATION_CMS_ID
            , 'Branch_Ownership__c'   => SETUP_BRANCH_OWNERSHIP
            , 'Status__c'             => 'Active'
            , 'Organisation_Email__c' => SETUP_ORGANISATION_EMAIL
            };
        sfab_FabricatedSObject fabAccGen = new sfab_FabricatedSObject( Account.class, accountGeneralMap );

        Map<string,object> subscriptionOrganisationMap = new Map<string,object>
            { 'Id'                    => 'Id-Sub-Org-1'
            , 'Organisation__c'       => fabAccGen
            , 'Publication__c'        => fabPubGen
            , 'Email_Subscription__c' => true
            , 'Status__c'             => 'Subscribed'
            , 'Email_Preference__c'   => 'Organisation Email'
            , 'Branch_Ownership__c'   => SETUP_BRANCH_OWNERSHIP
            , 'RecordTypeId'          => toRecordTypeId(Schema.SObjectType.Subscription__c, 'Organisation_Subscription')
            };
          sfab_FabricatedSObject fabSubOrg = new sfab_FabricatedSObject( Subscription__c.class, subscriptionOrganisationMap );

It is falling over on the last line.
Subscription is a custom object.  The field meta data for its Organisation field is:

    <?xml version="1.0" encoding="UTF-8"?>
    <CustomField xmlns="http://soap.sforce.com/2006/04/metadata">
        <fullName>Organisation__c</fullName>
        <deleteConstraint>Restrict</deleteConstraint>
        <externalId>false</externalId>
        <inlineHelpText>If the subscription is for an Organisation, select the name of the organisation here.</inlineHelpText>
        <label>Organisation</label>
        <lookupFilter>
            <active>true</active>
            <errorMessage>Please select a General, Ministry, Partner or Worker Account.</errorMessage>
            <filterItems>
                <field>Account.RecordType.Name</field>
                <operation>equals</operation>
                <value>General Account, Ministry Account, Partner Account, CMS Account</value>
            </filterItems>
            <isOptional>false</isOptional>
        </lookupFilter>
        <referenceTo>Account</referenceTo>
        <relationshipLabel>Subscriptions</relationshipLabel>
        <relationshipName>Subscriptions</relationshipName>
        <required>false</required>
        <type>Lookup</type>
    </CustomField>

The following code (which I was using prior to trying your library) works:

    private static Id toRecordTypeId(Schema.DescribeSObjectResult sObjDescribe, string recordTypeDevName) { return sObjDescribe.getRecordTypeInfosByDeveloperName().get(recordTypeDevName).getRecordTypeId(); }

        Publication__c generalPublication = new Publication__c 
            ( Name                = SETUP_NAME_PUBLICATION
            , RecordTypeId        = toRecordTypeId(Schema.SObjectType.Publication__c, 'General_Publication')
            , Branch_Ownership__c = SETUP_BRANCH_OWNERSHIP
            , Status__c           = SETUP_PUBLICATION_STATUS
            , Email__c            = true
            );
        insert generalPublication;

        Account generalAccount = new Account
            ( Name                  = SETUP_NAME_GENERAL_ACCOUNT
            , RecordTypeId          = toRecordTypeId(Schema.SObjectType.Account, 'General_Account')
            , CMS_ID__c             = SETUP_ORGANISATION_CMS_ID
            , Branch_Ownership__c   = SETUP_BRANCH_OWNERSHIP
            , Status__c             = 'Active'
            , Organisation_Email__c = SETUP_ORGANISATION_EMAIL
            );
        insert generalAccount;

        Publication__c generalPublication = [ SELECT Id, Name FROM Publication__c WHERE Name = :SETUP_NAME_PUBLICATION LIMIT 1 ];
        Account generalAccount = [ SELECT Id, Name FROM Account WHERE Name = :SETUP_NAME_GENERAL_ACCOUNT LIMIT 1 ];
        
        Subscription__c sub = new Subscription__c
            ( Organisation__c       = generalAccount.Id
            , Publication__c        = generalPublication.Id
            , Email_Subscription__c = true
            , Status__c             = 'Subscribed'
            , Email_Preference__c   = 'Organisation Email'
            , Branch_Ownership__c   = SETUP_BRANCH_OWNERSHIP
            , RecordTypeId          = toRecordTypeId(Schema.SObjectType.Subscription__c, 'Organisation_Subscription')
            );

        insert sub;

The same issue is also happening with 'Publication__c' => fabPubGen (if I remove 'Organisation__c' => fabAccGen)
I tried removing both the organisation and publication key values from the map and then doing;
    sfab_FabricatedSObject fabSubOrg = new sfab_FabricatedSObject( Subscription__c.class, subscriptionOrganisationMap );
    fabSubOrg.setParent(Subscription__c.Organisation__c, fabAccGen);
and also:
    fabSubOrg.setParent('Organisation__c', fabAccGen);
but these all yielded the same error.

Are you able to help?

Kind regards,
Craig.

**Comment by bobalicious** [comment]
Apologies - answering on my phone, rather than at a PC, so I may be off -
no ability to test.

Have you tried switching organisation __c over to organisation__r.
I.E. is the relationship and not the id that you're setting.

Also, I would consider moving over to using 'set' to set up the fields,
rather than the constructor.  As far as I am aware, the original intention
was for the constructor to be solely used when testing the library, rather
than for end-users. Though I guess that's down to personal taste - now the
API exists, it's unlikely to be removed.

**Comment by cdonnelly-cms** [comment]
Hi Rob,

Wow. Thank you for your quick response.

Using 'Organisation__r' seems to have fixed the problem.
That's a great help.

There's a new error now relating to an Id field being null, but it may be something to do with my test code, so I'll do some debugging tomorrow.

Thank you for the tip on using the set method.  I've switched over to:
new sfab_FabricatedSObject(Subscription__c.class).set(subscriptionOrganisationMap);

Kind regards,
Craig.

**Comment by cdonnelly-cms** [comment]
Hi again.

It looks like the null issue is occurring because the formula fields on the fabricated Subscription object are not populated.

Is there a way to get the formula fields to calculate or to manually populate them?

Kind regards,
Craig.

**Comment by cdonnelly-cms** [comment]
After some experimentation I found that setting the value of the formula fields directly worked.

Trying to get the formulas to calculate using...
        sfab_FabricatedSObject fabSubOrg = new sfab_FabricatedSObject(Subscription__c.class).set(subscriptionOrganisationMap);
        Subscription__c sub = (Subscription__c)fabSubOrg.toSObject();
        List<FormulaRecalcResult> results = Formula.recalculateFormulas(new List<Subscription__c> { sub });
        system.debug(results[0]);

Resulted in the following:

USER_DEBUG|[96]|DEBUG|FormulaRecalcResult:[recalcFieldErrors=(FormulaRecalcFieldError:[fieldError=Field not
supported: Publication__r. If this is lookup field, try setting the ID field instead., fieldName=Publication__r],
FormulaRecalcFieldError:[fieldError=Field not supported: Organisation__r. If this is lookup field, try setting the ID
field instead., fieldName=Organisation__r]), record=Subscription__c:{Is_Currently_Mailable_Subscription__c=false,
merge_Postcode__c=null, Address_Barcode__c=null, Postcode__c=null, Subscriber_Deceased__c=false,
mc_Is_Blank_Preferred_Email_Address__c=true, merge_State__c=null, mc_Is_Exclude_Contact_Individual__c=false,
Is_Mail_Address_Sufficient_for_Australia__c=false, Is_Subscription_for_Individual__c=false,
Email_Informal_Greeting__c=null, Individual_Lastname__c=null, Subscription_Record_Type_Name__c=Organisation Subscription,
State__c=null, Publication_Name__c=null, Key4_Individual_Publication_Email__c=&#124;&#124;1,
Publication_Record_Type_Name__c=null, merge_Street__c=null, Is_Mail_Country_Australia__c=true, Publication_Status__c=null,
Is_Mail_Address_Sufficient_for_Overseas__c=false, Email_Formal_Greeting__c=null, Email_Greeting__c=null,
temp_Is_Subscription_Email_Different__c=false, Email_Subscription__c=true, Email_Preference__c=Organisation Email,
merge_Is_Contact_Subscribed_and_Mailable__c=false, Id=a25000zzzz00001AAA, Status__c=Subscribed,
Mail_Informal_Greeting__c=null, Individual_Firstname__c=null, Email__c=null, Is_Subscribed__c=true, merge_Suburb__c=null,
Suburb__c=null, mc_Is_Sync_this_Record_Calc__c=false, merge_Organisation_Name__c=null, RecordTypeId=0126F000000o7jMQAQ,
Street__c=null, Subscription_Channels_Available__c=No delivery options have been set for this item,
merge_Mail_Formal_Greeting__c=null, Country__c=null, merge_Contact_Lastname__c=null, Monthly_Mailing_Segment__c=null,
Is_Mail_Street_Present__c=false, Organisation_Preferred_Name__c=null, mc_Subscriber_Key__c=null, Branch_Ownership__c=NSW &
ACT, Address_DPID__c=null, Address__c=null, merge_CMS_ID__c=null, Mail_Formal_Greeting__c=null, cc_Email__c=null,
mc_Is_Exclude_Contact_Person__c=false, mc_Is_Exclude_Organisation__c=true, CMS_ID__c=null, merge_Country__c=null,
Is_Subscription_for_Organisation__c=true, Individual_Full_Name__c= , mc_Is_Excluded__c=true}]

It looks like some of the formulas calculated but not the ones I needed.

**Comment by bobalicious** [comment]
I think this ticket is starting to get a little confused. If you need help
understanding a concept, or think there might be a bug, can you try to
reduce it to a much smaller illustration of the issue.  Generally 5 or 6
lines should do the job.

If this is the first time you've picked up the library, I suggest you try
it out with something simpler first and then build up to this.

Things I should point though:

* Formula fields are not calculated when building an object with sobject
fabricator.  Formulas are calculated when you retrieve data from the
database, which this library explicitly avoids.  Similar with roll-up
summaries (as far as usage is concerned).  If you need formulas to be set,
then set them directly - that's exactly the kind of thing this library
enables - setting read-only fields on sobjects.

* Custom relationships contain 2 fields:
    * ...__c contains the Id of the related object - you generally set that
when building things for insert and update, or retrieved if you want the Id
of a related record when you use SOQL.
    * ...__r contains the object that is related - it is generally set when
you retrieve relationship details with SOQL.
    * With sobject fabricator you are generally mimicking what happens when
you SOQL something, so you can use either as and when appropriate, by
setting either the __c or __r fields. The former with an Id, the latter
with another sobject fabricator instance.
    * You cannot set a __c field with an object, and you cannot set a __r
field with an Id.  The library will tell you when you get this wrong.

* Generally you should find that you don't need to build large scale
objects and relationships. You may find that your tests do not need all
this data to be set up.  Take a look at what your code / test needs, and
reduce what you are building to more accurately fit the needs.  Only write
the code you need to write.

In short though, I wish you luck, but I can't really support you with using
the library unless you can reduce your test cases to something
significantly smaller and ask a clear question / report a clearly described
bug (I did this, I expected it to do this, but it did this).

It's not that I don't want to, it's that it isn't really possible.

**Comment by cdonnelly-cms** [comment]
Thanks. All good now.

Sorry for providing too much detail. I'll try and be more succinct next time.

I understand better now the use case for this library in setting read only formula fields and the difference between setting the __c and __r fields.

My test is now working. Thank you for all your help.

**Comment by bobalicious** [comment]
No worries, glad I was of help.

Good luck with your future use!
